### PR TITLE
feat(ui): upgrade RMWC

### PR DESCRIPTION
**Pull request by endymion1818**
## Changes
- Upgraded RMWC components to v7
- Upgrade @material components to versions used in v7 of RMWC
- Import SASS to use new syntax
- Update styles to resolve changes from the upgrade

## How can this be tested?
- https://d3mmrkoy170qpm.cloudfront.net/

### CODE REVIEW

File: apps/theme/pageBuilder/styles/medium.scss [hunk 1]
```diff
@@ -1,8 +1,8 @@
 /**
  *  `Medium editor` styles
  */
-@import "~medium-editor/dist/css/medium-editor.css";
-@import "~medium-editor/dist/css/themes/mani.css";
+@import "medium-editor/dist/css/medium-editor.css";
```
Comment (Pavel910) [inline apps/theme/pageBuilder/styles/medium.scss:4]: Existing user projects will still have imports via `~`. Will this upgrade break this for those projects, or is this backwards compatible? This has to be tested with older projects.

File: apps/theme/pageBuilder/styles/medium.scss [hunk 2]
```diff
@@ -1,8 +1,8 @@
...
+@import "medium-editor/dist/css/themes/mani.css";
```
Comment (Pavel910) [inline apps/theme/pageBuilder/styles/medium.scss:5]: `mani.css` ? I have a feeling this should be `main.css`. Please double-check.

File: apps/theme/styles.scss [hunk 3]
```diff
@@ -1,3 +1,3 @@
 @use "./pageBuilder/styles/theme.scss";
 @use "./formBuilder/styles/theme.scss" as themeFormBuilder;
-@use "~@webiny/react-rich-text-renderer/styles.scss";
+@import "@webiny/react-rich-text-renderer/styles.scss";
```
Comment (Pavel910) [inline apps/theme/styles.scss:3]: What's the difference between `@use` and `@import`, and why is `import` only used  on the last scss file, but not the first two as well?

File: package.json [hunk 4]
```diff
@@ -102,6 +102,7 @@
     "rxjs": "^6.5.5",
     "semver": "^6.3.0",
     "ts-jest": "^26.4.4",
+    "tslib": "2.4.0",
```
Comment (Pavel910) [inline package.json]: Why do we need this `tslib` package? We didn't need it thus far, I want to know what dependency made this a requirement, and if it has to sit in the root package.json.

File: package.json [hunk 5]
```diff
@@ -211,20 +212,41 @@
...
+    "@material/form-field": "^5.1.0",
+    "@material/icon-button": "^5.1.0",
+    "@material/linear-ripple": "^5.1.0",
+    "@material/list": "^5.1.0",
+    "@material/radio": "^5.1.0",
+    "@material/select": "^5.1.0",
+    "@material/snackbar": "^5.1.0",
+    "@material/switch": "^5.1.0",
+    "@material/tab-scroller": "^5.1.0",
+    "@material/top-app-bar": "^5.1.0"
```
Comment (Pavel910) [inline package.json]: This is the wrong place for these dependencies. They should be moved to the package that's actually using these. If we're using RMWC, I'm not sure why we need to import these as dependencies at all. We can discuss this over a call if you're unsure what to do with these.

File: packages/app-admin/src/components/AppInstaller/styled.tsx [hunk 6]
```diff
@@ -39,11 +39,12 @@ export const InstallContent = styled("div")({
...
-        height: "100vh",
-        ".webiny-split-view__right-panel-wrapper": {
-            height: "100vh"
-        }
+        height: "100vh"
+    },
+    ".webiny-split-view__right-panel-wrapper": {
+        height: "100vh"
     }
 });
```
Comment (Pavel910) [inline packages/app-admin/src/components/AppInstaller/styled.tsx:49]: If I remember this correctly, this `100vh` should apply to both left and right panels. With this new rule it looks like you're only applying that height to the right panel.

File: packages/app-form-builder/package.json [hunk 7]
```diff
@@ -19,7 +19,7 @@
     "@editorjs/editorjs": "^2.20.1",
     "@emotion/core": "^10.0.17",
     "@emotion/styled": "^10.0.17",
-    "@rmwc/menu": "^5.6.0",
+    "@rmwc/menu": "7.0.3",
```
Comment (Pavel910) [inline packages/app-form-builder/package.json]: This bothers me because `@rmwc` packages should only be used in the `ui` package. Please make a note of these packages, and we should solve this in a separate PR. All our apps should be using the `ui` package exclusively, to maintain the abstraction in one place.

File: packages/app-page-builder/package.json [hunk 8]
```diff
@@ -21,7 +21,7 @@
     "@fortawesome/free-regular-svg-icons": "^6.0.0",
     "@fortawesome/free-solid-svg-icons": "^6.0.0",
     "@fortawesome/react-fontawesome": "^0.1.17",
-    "@rmwc/menu": "^5.6.0",
+    "@rmwc/menu": "7.0.3",
```
Comment (Pavel910) [inline packages/app-page-builder/package.json]: Same as the previous file.

File: packages/cwp-template-aws/template/common/apps/theme/styles.scss [hunk 9]
```diff
@@ -1,3 +1,3 @@
 @import "./pageBuilder/styles/theme.scss";
 @import "./formBuilder/styles/theme.scss";
-@import "~@webiny/react-rich-text-renderer/styles.scss";
+@import "Z@webiny/react-rich-text-renderer/styles.scss";
```
Comment (Pavel910) [inline packages/cwp-template-aws/template/common/apps/theme/styles.scss]: A typo: `Z@webiny` :)

File: packages/project-utils/bundling/app/config/webpack.config.js [hunk 10]
```diff
@@ -40,9 +40,11 @@ const cssRegex = /\.css$/;
 const cssModuleRegex = /\.module\.css$/;
 const sassRegex = /\.(scss|sass)$/;
 const sassModuleRegex = /\.module\.(scss|sass)$/;
+
 const sassLoader = {
     loader: require.resolve("sass-loader"),
     options: {
+        // webpackImporter: false,
```
Comment (Pavel910) [inline packages/project-utils/bundling/app/config/webpack.config.js]: Leave a comment why this is commented out.

File: packages/ui/src/Chips/ChipIcon.tsx [hunk 11]
```diff
@@ -1,5 +1,5 @@
 import React from "react";
-import { ChipIcon as RmwcChipIcon, ChipIconProps as RmwcChipIconProps } from "@rmwc/chip";
+import { Chip as RmwcChipIcon, ChipProps as RmwcChipIconProps } from "@rmwc/chip";
```
Comment (Pavel910) [inline packages/ui/src/Chips/ChipIcon.tsx]: Previously, there were 2 separate components: a `Chip`, and a `ChipIcon`. `ChipIcon` was used as a child of the `Chip` component to add an icon to a chip, here's an example:

```
<Chip>
    <ChipIcon icon={<AnIcon/>}/>
    Chip Label
</Chip>
```

This was changed ages ago, but this `<ChipIcon>` component was not removed. We don't use it anywhere in Webiny, so feel free to delete this component and stories for it. It's meaningless because the `<Chip>` itself supports `icon` props.

**Review by Pavel910 — CHANGES_REQUESTED** [review]
@endymion1818 did the first review so you can get some things sorted before jumping on a call. Let me know if any of my comments need clarification.

File: apps/theme/pageBuilder/styles/medium.scss (hunk 2 above)
Comment (endymion1818) [reply apps/theme/pageBuilder/styles/medium.scss:5]: `mani` is in fact the name of the Medium editor theme we're using: https://github.com/yabwe/medium-editor/blob/master/src/sass/themes/mani.scss 😆

File: package.json (hunk 4 above)
Comment (endymion1818) [reply package.json]: it's an implicit dependency of a few projects, and as a result of different versions being used you get this error on building:

`admin: Error: Attempted import error: '__spreadArray' is not exported from 'tslib' (imported as '__spreadArray').`

Unfortunately you still get this error if it's only listed in the `resolutions`.

File: apps/theme/styles.scss (hunk 3 above)
Comment (endymion1818) [reply apps/theme/styles.scss:3]: ~We bumped up `sass` and `sass-loader` as part of this upgrade:~

- `@import` globally imports other sass files and [is being phased out](https://sass-lang.com/documentation/at-rules/import). It's appropriate here because we're importing `rich-text-renderer` styles from another package, whereas we're using the other two that come from this package. Also you get errors any other way around 🙃
- `@use` forwards mixins, variables etc _to the current sass file_, so it's more explicit as to what's happening
- `~` (tilde) is deprecated from `sass-loader` (https://webpack.js.org/loaders/sass-loader/#resolving-import-at-rules)

File: package.json (hunk 5 above)
Comment (endymion1818) [reply package.json]: I moved these to `/packages/ui/package.json` and the build was successful. Sorry, my mistake I thought resolutions had to be at the top level `package.json` because the dependency `material-components-web` which also uses them is in this `package.json`.

File: packages/app-admin/src/components/AppInstaller/styled.tsx (hunk 6 above)
Comment (endymion1818) [reply packages/app-admin/src/components/AppInstaller/styled.tsx:49]: This is just a syntax change, I've removed the nesting from the declaration as it wasn't getting evaluated.

File: packages/app-form-builder/package.json (hunk 7 above)
Comment (endymion1818) [reply packages/app-form-builder/package.json]: Yep makes perfect sense

File: packages/app-page-builder/package.json (hunk 8 above)
Comment (endymion1818) [reply packages/app-page-builder/package.json]: 👍🏻

File: packages/cwp-template-aws/template/common/apps/theme/styles.scss (hunk 9 above)
Comment (endymion1818) [reply packages/cwp-template-aws/template/common/apps/theme/styles.scss]: 🙈

File: packages/project-utils/bundling/app/config/webpack.config.js (hunk 10 above)
Comment (endymion1818) [reply packages/project-utils/bundling/app/config/webpack.config.js]: It should be removed, I was doing a test to see if it materially changed anything but it didn't

File: packages/ui/src/Chips/ChipIcon.tsx (hunk 11 above)
Comment (endymion1818) [reply packages/ui/src/Chips/ChipIcon.tsx]: Done, thanks

File: apps/theme/pageBuilder/styles/medium.scss (hunk 1 above)
Comment (endymion1818) [reply apps/theme/pageBuilder/styles/medium.scss:4]: It's being deprecated but I couldn't find any notification as to when they might remove it. https://webpack.js.org/loaders/sass-loader/#resolving-import-at-rules

File: packages/ui/package.json [hunk 12]
```diff
@@ -116,6 +116,30 @@
...
+    "@material/icon-button": "^5.1.0",
+    "@material/linear-ripple": "^5.1.0",
+    "@material/list": "^5.1.0",
+    "@material/radio": "^5.1.0",
+    "@material/select": "^5.1.0",
+    "@material/snackbar": "^5.1.0",
+    "@material/switch": "^5.1.0",
+    "@material/tab-scroller": "^5.1.0",
+    "@material/top-app-bar": "^5.1.0"
+  },
```
Comment (Pavel910) [inline packages/ui/package.json]: You cannot have `resolutions` per workspace. `resolutions` is only valid in the monorepo root, as their purpose is to resolve to one common dependency version for the entire project. And yarn tells you that when you try to install deps:

CleanShot 2022-05-23 at 12 58 57@2x

My main question is: why do you need to specify these at all? Aren't these packages internal dependencies of @rmwc packages?

From what I see, @rmwc has these as explicit dependencies in their packages:
CleanShot 2022-05-23 at 13 01 39@2x

This means that you don't need to specify any `@material/*` packages as Webiny dependencies at all.

Comment (endymion1818) [reply packages/ui/package.json]: Thanks for the convo about this, the issue is resolved by removing these dependencies. It turns out that updating the `material-components-web` package made it redundant.

**Comment by Pavel910** [comment]
Found this weird behavior on Pages list filters:

https://user-images.githubusercontent.com/3920893/171034441-34a96d71-742e-4452-aab6-4bd7bbdea1bc.mp4

However, this works fine on Content Entries list filtering. Probably `select` options are not passed in the same way. RMWC 5 had some weird stuff going on with select boxes.

**Comment by Pavel910** [comment]
Another thing to check, something related to Portals:

CleanShot 2022-05-30 at 19 32 33@2x

Add that `<Portal/>` component here: `packages/app-admin-rmwc/src/modules/Overlays/index.tsx:19`. There's a comment in that file describing what that stuff does.

**Comment by Pavel910** [comment]
Secondary button styles are slightly off, this is how they currently are, and with the new RMWC, the border is different.

CleanShot 2022-05-30 at 22 57 22@2x

**Comment by Pavel910** [comment]
@endymion1818 Adding more screenshots to resolve:

The following three screenshots show the same issue, the floating label is positioned a bit off. Seems like it's only an issue with values prefilled by the browser. RMWC is probably not detecting that there's a value in the field (just guessing here).

CleanShot 2022-06-05 at 16 01 35

CleanShot 2022-06-05 at 16 02 34

CleanShot 2022-06-05 at 16 04 28

Page Builder glitches:

CleanShot 2022-06-05 at 16 05 39

Menu doesn't seem to work at all (that's probably that thing with Portal):

https://user-images.githubusercontent.com/3920893/172054602-132654e5-8cad-431a-91af-e9da1a4d4fdb.mp4

These buttons are off:

CleanShot 2022-06-05 at 16 08 43

File: apps/admin/package.json [hunk 13]
```diff
@@ -28,7 +28,8 @@
     "react": "16.14.0",
     "react-dom": "16.14.0",
     "regenerator-runtime": "^0.13.5",
-    "theme": "^0.1.0"
+    "theme": "^0.1.0",
+    "tslib": "^2.4.0"
```
Comment (Pavel910) [inline apps/admin/package.json:32]: @endymion1818 just to confirm: is `tslib` dependency still necessary? I know you mentioned that it's our rmwc forks that were causing issues, is that related?

File: packages/ui/src/Button/Button.styles.ts [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+ * @type {string}
+ */
+const webinyButtonStyles = css(
+    {},
+    {
+        "&.webiny-ui-button--secondary:not(:disabled)": {
+            borderColor: "var(--mdc-theme-primary)"
+        }
+    }
+);
```
Comment (Pavel910) [inline packages/ui/src/Button/Button.styles.ts]: @endymion1818 This will look much nicer if you use the template literals; then you also don't need to use the camelCase for properties. You just write plain CSS instead.
```
const webinyButtonStyles = css`
    &.webiny-ui-button--secondary:not(:disabled) {
        border-color: var(--mdc-theme-primary)
    }
`
```

File: packages/ui/src/Input/Input.tsx [hunk 15]
```diff
@@ -37,8 +39,24 @@ export type InputProps = FormComponentProps &
...
+ * @type {string}
+ */
+const webinyInputStyles = css(
+    {},
+    {
+        ".mdc-text-field__input:-webkit-autofill + .mdc-floating-label": {
+            transform: "translateY(-106%) scale(0.75)"
+        }
+    }
+);
```
Comment (Pavel910) [inline packages/ui/src/Input/Input.tsx]: Same as previous comment; try using template literals.

**Review by Pavel910 — CHANGES_REQUESTED** [review]
@endymion1818 just a few minor things and we're good with the code review. The rest looks pretty solid. It's mostly adding of `data-testid` and some CSS tweaks across the board, so there isn't a whole lot of code changes that can break anything.

File: apps/admin/package.json (hunk 13 above)
Comment (endymion1818) [reply apps/admin/package.json:32]: @Pavel910 unfortunately yes, Im afraid we still get that issue with the `spreadArray` function we identified before without explicitly adding this line.

File: packages/ui/src/Button/Button.styles.ts (hunk 14 above)
Comment (endymion1818) [reply packages/ui/src/Button/Button.styles.ts]: I was following what I thought was a convention but very happy you said this, I prefer template literals too.

File: apps/admin/package.json (hunk 13 above)
Comment (Pavel910) [reply apps/admin/package.json:32]: In that case, I assume we'll need to add this dependency as part of the project upgrade we're preparing for 5.30. Hopefully it's the only thing that needs changing in existing user projects. Have you tried upgrading from 5.29 ?

Comment (endymion1818) [reply apps/admin/package.json:32]: Ah, no haven't tested that. I'll run through it in the morning.
